### ISSUE TITLE
Refactor the library

### ISSUE THREAD

**Issue by DigitalBrainJS** [EventEmitter2/EventEmitter2#257]
The library has strange, almost dead conditional branches with checking `_events` and `_all` props, since _events is always truly, moreover it is initialized with object even in wildcard mode, but never used.
https://github.com/EventEmitter2/EventEmitter2/blob/9f8a8371df3088469d02992aba244bcfc1eeb832/lib/eventemitter2.js#L222
The _all array stay initialized for always after being first time used.
So in fact these branches are dead (never executed at all):
https://github.com/EventEmitter2/EventEmitter2/blob/9f8a8371df3088469d02992aba244bcfc1eeb832/lib/eventemitter2.js#L459
https://github.com/EventEmitter2/EventEmitter2/blob/9f8a8371df3088469d02992aba244bcfc1eeb832/lib/eventemitter2.js#L567
These checks are useless:
https://github.com/EventEmitter2/EventEmitter2/blob/9f8a8371df3088469d02992aba244bcfc1eeb832/lib/eventemitter2.js#L712
https://github.com/EventEmitter2/EventEmitter2/blob/9f8a8371df3088469d02992aba244bcfc1eeb832/lib/eventemitter2.js#L463
https://github.com/EventEmitter2/EventEmitter2/blob/9f8a8371df3088469d02992aba244bcfc1eeb832/lib/eventemitter2.js#L876
https://github.com/EventEmitter2/EventEmitter2/blob/9f8a8371df3088469d02992aba244bcfc1eeb832/lib/eventemitter2.js#L904
And so on.
It seems we should set _events and _all to null if there are no any listeners, or use internal flag to determine if there are any listeners to avoid object reinitiazlization.

`emitter.emitAsync` has bad logic and returns very strange values.
If _all and _events was not initialized yet it returns **false** (but fortunately never executed) not even Promise.resolve(false).
For `newListener` event and if the vent has no listeners it returns [false]
If _all and _events is initialized and there are no any listeners - []
Moreover it's resolves promises from `all` listeners array and array of specific event type so there is no ability to determine is value returned from `all` listener or from a listener of specific event type. 
**I bellive this should be refactored in major version:**
- The method should return an empty array or false if there are no methods listening the specific event.
- The method should not collect the values from `all` listeners, just wait while they will be resolved.

**Comment by RangerMauve** [comment]
Yeah, I don't think I was around when that was introduced. I think emitAsync was a PR which I didn't look too deeply into.

I'm down for a major bump. 😁

**Comment by DigitalBrainJS** [comment]
Yeap. This lib has a bit more issues that just emitAsync and currently i have not an idea how to fix that and keep 100% compability. 
Some internal props isn't marked as private for ex. this.listenerTree, this.delimiter, this.wildcard so i can't refactor that and be sure nobody uses them.
this.listenerTree should be refactored to this._events to simplify code logic
this.delimiter should be refactored to this._delimiter 
this.wildcard should be refactored to this._wildcard and so on. In the future their keys could be replaced with private or symbol variables.
this.emit always returns true, even if all the listeners have been removed...

**Comment by DigitalBrainJS** [comment]
[It doesn’t look so good :) We have to fix this ;)](https://www.npmtrends.com/eventemitter2-vs-eventemitter3-vs-eventemitter4-vs-events-vs-EventEmitter)

**Comment by DigitalBrainJS** [comment]
@RangerMauve I just found out that wildcard emitter benchmark is totally fake 😄 
https://github.com/EventEmitter2/EventEmitter2/blob/f3f642f17e081ec605e4504b8c5ca002d3f663ba/test/perf/benchmark.js#L11
https://github.com/EventEmitter2/EventEmitter2/blob/799d2f83305ce6752d54038e865416942763ed64/test/perf/benchmark.js#L38-L44
Previously i wonder how that heavy code has a performance only three times less than highly optimized plain emitter. Event plain namespace.split('.') has performance only around 3M op/sec))
So, the secret is revealed. That benchmark tests an emitter not in wildcard mode and the real performance is near 300k operations/sec, not 3.5 millions :D It's 40-50 times slower than simply emitter has.

**Comment by RangerMauve** [comment]
Ah dang! I wonder how long it's been that way. 😂

**Comment by DigitalBrainJS** [comment]
@RangerMauve not for long, It looks like just about 9 years :)
It seems searchListenerTree is too slow and over complicated for emit purpose.
And it's hard to refactor it, since I don’t know fully how it should work with double ** wildcard and README has almost no info about that. Is unlikely that someone will be use wildcard if it is 50 times slower than simple emitter...

**Comment by RangerMauve** [comment]
When I was using this library actively at work (years ago) I exclusively used it for the wildcards. 😂

**Comment by DigitalBrainJS** [comment]
@RangerMauve Hm. I never needed to use wildcards emitters, at least something harder than basic one level wildcard e.g. `event.*` . This lib supports multilevel * and ** and in fact all logic depends on one complex function with heavy recursion that has a pretty low performance. I can't image a challange where we need namespace like `event.**.foo.bar`.
The previos benchmark was a benchmark for a wildcard emitter from git branch with some optimization which affected the results. The reality a bit worse (origin master branch):

> EventEmitter2 x 12,089,852 ops/sec ±4.00% (62 runs sampled)
> **EventEmitter2 (wild) x 167,847 ops/sec ±4.90% (63 runs sampled)**
> EventEmitter2 (wild) using plain events x 190,736 ops/sec ±3.08% (62 runs sampled)
> EventEmitter2 (wild) emitting ns x 676,378 ops/sec ±4.07% (67 runs sampled)
> EventEmitter2 (wild) emitting a plain event x 764,142 ops/sec ±3.87% (71 runs sampled)
> EventEmitter3 x 10,223,124 ops/sec ±3.51% (70 runs sampled)

Even in a case of emitting a simple wildcard event (test.*) it's 75 times slower than a general emitter.
